adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,117 @@
+= Enhanced extension metadata generation
+
+* Status: proposed
+* Date: 2024-04-26
+
+== Context and Problem Statement
+
+How do we know which extensions provide a dev service? At the moment, we don't, except by folk knowledge and https://quarkus.io/guides/dev-services[manually maintained lists.]
+
+For further context on the need, see also
+
+* https://github.com/quarkusio/quarkus/pull/40306
+* https://github.com/quarkusio/quarkusio.github.io/issues/1869
+* https://github.com/quarkusio/extensions/issues/983
+
+When building an extension, we have an opportunity to introspect and surface a range of useful metadata, such as whether it provides a dev service.
+But there's a problem. The introspection needs to happen in the deployment module, because that's where most of the interesting extension code is.
+But the place we create the `quarkus-extension.yaml` is the runtime module. The deployment module depends on the runtime module, so it's a circular dependency.
+
+Because the dependency isn't a code-compilation one, it doesn't break the build, but it does mean the generated metadata is only accurate on the second build.
+
+image::images/0002-circular-dependency.svg[the circular dependency]
+Source: https://excalidraw.com/#json=cTOOltCdWxadxC8QYc93R,AzxQWd7ALVSsk_FbyIuqAg
+
+=== Related problems
+
+We have a similar circularity problem with config javadoc; the config is discovered in the deployment module, but the generated of the final-format data is done elsewhere. A solution here *might* help with that.
+
+Although it's not a serious problem, it would also be nice to allow people to use `CapabilityBuildItems` to register capabilities.
+At the moment, if authors do that, the information https://quarkus.io/guides/capabilities#capabilitybuilditem[will not be available to the dev tools], so we need to have lots of warnings in our documentation telling them not to.
```

Comment:
I added a comment about it https://github.com/quarkusio/quarkus/pull/40306#issuecomment-2235844459 too.
We may want to deprecate `CapabilityBuildItem` in the future and prefer the metadata-only approach.

Reply:
Yes, I can see the appeal of that. I'd say that we should either get rid of it, or make it work properly without warnings not to use it. :)